gapis/api/transform2/capture_log.go

```diff
@@ -0,0 +1,130 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package transform2
+
+import (
+	"context"
+	"os"
+
+	"github.com/google/gapid/core/log"
+	"github.com/google/gapid/core/memory/arena"
+	"github.com/google/gapid/gapis/api"
+	"github.com/google/gapid/gapis/capture"
+)
+
+type captureLog struct {
+	file   *os.File
+	header *capture.Header
+	cmds   []api.Cmd
+	ids    map[api.CmdID]api.CmdID
+}
+
+// NewCaptureLog returns a Transformer that will log all commands passed through it
+// to the capture file at path.
+func NewCaptureLog(ctx context.Context, sourceCapture *capture.GraphicsCapture, path string) *captureLog {
+	f, err := os.Create(path)
+	if err != nil {
+		log.E(ctx, "Failed to create replay capture file %v: %v", path, err)
+		return nil
+	}
+	return &captureLog{
+		file:   f,
+		header: sourceCapture.Header,
+		cmds:   []api.Cmd{},
+		ids:    map[api.CmdID]api.CmdID{},
+	}
+}
+
+func (logTransform *captureLog) RequiresAccurateState() bool {
+	return false
+}
+
+func (logTransform *captureLog) BeginTransform(ctx context.Context, inputCommands []api.Cmd, inputState *api.GlobalState) ([]api.Cmd, error) {
+	return inputCommands, nil
+}
+
+func (logTransform *captureLog) ClearTransformResources(ctx context.Context) {
+	// Do nothing
+}
+
+func (logTransform *captureLog) TransformCommand(ctx context.Context, id api.CmdID, inputCommands []api.Cmd, inputState *api.GlobalState) ([]api.Cmd, error) {
+	// Melih TODO: We cannot have this check anymore.
+	// We can make an approximation that first command in the array likely to be a real command
+	// but it's still may not be true
+
+	// Don't write out the custom commands (e.g. replay.Custom)
+	// if cmd.API() != nil {
+	// 	if id.IsReal() {
+	// 		t.ids[id] = api.CmdID(len(t.cmds))
+	// 	}
+	// 	t.cmds = append(t.cmds, cmd)
+	// }
+
+	if len(inputCommands) > 0 {
+		logTransform.ids[id] = api.CmdID(len(logTransform.cmds))
+		logTransform.cmds = append(logTransform.cmds, inputCommands[0])
+	}
+
+	return inputCommands, nil
+}
+
+func (logTransform *captureLog) EndTransform(ctx context.Context, inputCommands []api.Cmd, inputState *api.GlobalState) ([]api.Cmd, error) {
+	a := arena.New()
+	defer a.Dispose()
+
+	for idx := range logTransform.cmds {
+		cmd := logTransform.cmds[idx].Clone(a)
+		logTransform.cmds[idx] = cmd
+
+		// Melih TODO: We cannot check this anymore
+		// Due to this issue, probably functionality is broken
+
+		// if caller := cmd.Caller(); caller.IsReal() {
```

Comment:
See note above in this file

Reply:
self note: Try to remove caller mechanism completely. It's OpenGL related.